src/SortingAlgorithms.jl

```diff
@@ -652,4 +681,239 @@ else
+            while use_a ? a <= m - pagesize + 1 : b <= hi - pagesize + 1
+                pages = next_page!(pageLocations, pages, pagesize, lo, a)
+                offset = page_offset(pages.current)
+                copyto!(v, offset + 1, v, use_a ? a : b, pagesize)
+                use_a && (a += pagesize)
+                use_a || (b += pagesize)
+            end
+            # copy the final partial page only if sourcing from A.
+            # If sourcing from B, it is already in place.
+            use_a && copyto!(v, hi - m + a, v, a, m - a + 1)
+        end
+
+        ##################
+        # rearrange pages
+        ##################
+        # copy pages belonging to the 3 permutation chains ending with a page in the scratch space
+        nextA, nextB = pages.nextA, pages.nextB
+
+        for _ in 1:3
+            page = (nextB > nPages ? (nextA += 1) : (nextB += 1)) - 1
+            page = permute_pages!(>(3), v, pageLocations, page_offset, pagesize, page)
+            copyto!(v, page_offset(page) + 1, scratch, (page - 1) * pagesize + 1, pagesize)
+        end
+
+        # copy remaining permutation cycles
+        for donePageIndex = 5:nPages
+            # linear scan through pageLocations to make sure no cycle is missed
+            page = pageLocations[donePageIndex-3]
+            page == donePageIndex && continue
+
+            # copy the data belonging to donePageIndex into scratch
+            copyto!(scratch, 1, v, page_offset(page) + 1, pagesize)
+
+            # follow the cycle starting with the newly freed page
+            permute_pages!(!=(donePageIndex), v, pageLocations, page_offset, pagesize, page)
+            copyto!(v, page_offset(donePageIndex) + 1, scratch, 1, pagesize)
+        end
+    end
+end
+
+# midpoint was added to Base.sort in version 1.4 and later moved to Base
+# -> redefine for compatibility with earlier versions
+midpoint(lo::Integer, hi::Integer) = lo + ((hi - lo) >>> 0x01)
+
+function pagedmergesort!(v::AbstractVector{T}, lo::Integer, hi::Integer, o::Ordering, scratch::AbstractVector{T}, pageLocations) where {T}
+    len = hi + 1 - lo
+    if len <= Base.SMALL_THRESHOLD
+        return Base.Sort.sort!(v, lo, hi, Base.Sort.InsertionSortAlg(), o)
+    end
+    m = midpoint(lo, hi - 1) # hi-1: ensure midpoint is rounded down. OK, because lo < hi is satisfied here
+    pagedmergesort!(v, lo, m, o, scratch, pageLocations)
+    pagedmergesort!(v, m + 1, hi, o, scratch, pageLocations)
+    if len <= length(scratch)
+        twoended_merge!(v, lo, m, hi, o, scratch)
+    else
+        paged_merge!(v, lo, m, hi, o, scratch, pageLocations)
+    end
+    return v
+end
+
+function sort!(v::AbstractVector, lo::Integer, hi::Integer, ::PagedMergeSortAlg, o::Ordering)
+    lo >= hi && return v
+    n = hi + 1 - lo
+    pagesize = isqrt(n)
+    scratch = Vector{eltype(v)}(undef, 3pagesize)
+    nPages = n ÷ pagesize
+    pageLocations = Vector{Int}(undef, nPages - 3)
```

Comment:
If `nPages < 3`, this will throw. This happens when `2 <= length(lo:hi) <= 5`.

Reply:
Fixed in [717f9ae](https://github.com/JuliaCollections/SortingAlgorithms.jl/pull/71/commits/717f9ae6273c3e14136354eae3c4cc4c09e26629).
Do you think an extra test is needed for this?
(i.e. test SortingAlgorithms.PagedMergeSortAlg() in addition to PagedMergeSort)

Reply:
CI revealed this bug, so it's not totally necessary, but it certainly wouldn't hurt.

Reply:
Test implemented in [d367ee8](https://github.com/JuliaCollections/SortingAlgorithms.jl/pull/71/commits/d367ee8a17a087c6c54994e9375269782e8334dd).